File: src/core/EffectiveAddress.cpp
#include "core/EffectiveAddress.hpp"

#include "core/Cpu.hpp"
#include "utils/log.hpp"

using namespace rbt;

[[nodiscard]] std::optional<IndexExtension> IndexExtension::decode(
	const Mmu& mmu, u32 pc
) {
	auto data = mmu.read_be16(pc);
	if (!data) {
		log::warn(
			"[CPU]({:#010x}) > Failed to read brief extension word: {}", pc, data.error()
		);
		return std::nullopt;
	}

	return IndexExtension::decode(*data);
}

[[nodiscard]] std::optional<IndexExtension> IndexExtension::decode(u16 ext) {
	if ((ext >> 8) & 0x00) {
		log::warn("[CPU] > M68010 doesn't support full extension word!");
		return std::nullopt;
	}

	// Brief Extension Word
	// | F | E  | D  | C  | B | A | 9 | 8 | 7 | 6 | 5 | 4 | 3 | 2 | 1 | 0 |
	// |A/D| Register     | S | Scale | 0 | Displacement Integer          |
	IndexExtension ix {};
	ix.is_addr = (ext >> 15) & 0x01; // A/D: 0=Dn; 1=An
	ix.is_long = (ext >> 11) & 0x01; // S: 0=W; 1=L
	ix.reg = (ext >> 12) & 0x07;
	ix.scale = (ext >> 9) & 0x03;
	ix.displacement = operand_sign_extend(OperandSize::Byte, ext & 0xff);

	return ix;
}

u16 IndexExtension::encode() const {
	u16 ext = 0;

	ext |= (is_addr & 0x01) << 15;
	ext |= (is_long & 0x01) << 11;
	ext |= (reg & 0x07) << 12;
	ext |= (scale & 0x03) << 9;
	ext |= (displacement & 0xff);

	return ext;
}

[[nodiscard]] std::optional<EffectiveAddress> EffectiveAddress::decode(
	u8 mode, u8 reg, u8 size, const Mmu& mmu, u32 pc
) {
	EffectiveAddress ea {};
	ea.reg = reg;
	ea.size = OperandSize::None;
	ea.program_counter = pc;

	switch (mode) {
	case 0b000: // Dn
		ea.mode = AddressMode::DirectData;
		ea.reg_type = RegisterType::Data;
		ea.size = static_cast<OperandSize>(size);

		break;
	case 0b001: // An
		ea.mode = AddressMode::DirectAddress;
		ea.reg_type = RegisterType::Address;
		ea.size = static_cast<OperandSize>(size);

		break;
	case 0b010: // (An)
		ea.mode = AddressMode::Indirect;
		ea.reg_type = RegisterType::Address;

		break;
	case 0b011: // (An)+
		ea.mode = AddressMode::IndirectPostInc;
		ea.reg_type = RegisterType::Address;
		ea.size = static_cast<OperandSize>(size);

		break;
	case 0b100: // -(An)
		ea.mode = AddressMode::IndirectPreDec;
		ea.reg_type = RegisterType::Address;
		ea.size = static_cast<OperandSize>(size);

		break;
	case 0b101: { // (d16, An)
		ea.mode = AddressMode::IndirectDisplacement;
		ea.reg_type = RegisterType::Address;
		ea.bytes_read = 2;

		auto disp = mmu.read_be16(pc);
		if (!disp) {
			log::warn(
				"[CPU]({:#010x}) > Failed to read absolute address: {}", pc, disp.error()
			);
			return std::nullopt;
		}

		ea.displacement = operand_sign_extend(OperandSize::Word, *disp);
	} break;
	case 0b110: // (d8, An, Xn)
		ea.mode = AddressMode::IndirectIndex;
		ea.reg_type = RegisterType::Address;
		ea.bytes_read = 2;
		ea.index = IndexExtension::decode(mmu, pc);
		if (!ea.index) {
			return std::nullopt;
		}

		break;
	case 0b111:
		switch (reg) {
		case 0b000: { // (xxx).W
			ea.mode = AddressMode::AbsoluteShort;
			ea.reg_type = RegisterType::None;
			ea.bytes_read = 2;

			auto abs = mmu.read_be16(pc);
			if (!abs) {
				log::warn(
					"[CPU]({:#010x}) > Failed to read absolute address: {}", pc,
					abs.error()
				);
				return std::nullopt;
			}

			ea.absolute = *abs;
		} break;
		case 0b001: { // (xxx).L
			ea.mode = AddressMode::AbsoluteLong;
			ea.reg_type = RegisterType::None;
			ea.bytes_read = 4;

			auto abs = mmu.read_be32(pc);
			if (!abs) {
				log::warn(
					"[CPU]({:#010x}) > Failed to read absolute address: {}", pc,
					abs.error()
				);
				return std::nullopt;
			}

			ea.absolute = *abs;
		} break;
		case 0b010: { // (d16, PC)
			ea.mode = AddressMode::ProgramCounterDisplacement;
			ea.reg_type = RegisterType::ProgramCounter;
			ea.bytes_read = 2;

			auto disp = mmu.read_be16(pc);
			if (!disp) {
				log::warn(
					"[CPU]({:#010x}) > Failed to read displacement: {}", pc, disp.error()
				);
				return std::nullopt;
			}

			ea.displacement = operand_sign_extend(OperandSize::Word, *disp);
		} break;
		case 0b011: // (d8, PC, Xn)
			ea.mode = AddressMode::ProgramCounterIndex;
			ea.reg_type = RegisterType::ProgramCounter;
			ea.bytes_read = 2;
			ea.index = IndexExtension::decode(mmu, pc);
			if (!ea.index) {
				return std::nullopt;
			}

			break;
		case 0b100: { // #imm
			if (size >= 0x03) {
				log::warn(
					"[CPU] > Invalid address size for immediate mode: {:#05b}", size
				);
				return std::nullopt;
			}

			ea.mode = AddressMode::Immediate;
			ea.reg_type = RegisterType::None;
			ea.size = static_cast<OperandSize>(size);

			auto imm = mmu.load(pc, ea.size);
			if (!imm) {
				log::warn(
					"[CPU]({:#010x}) > Failed to read immediate data: {}", pc, imm.error()
				);
				return std::nullopt;
			}

			ea.immediate = *imm;
			ea.bytes_read = ea.size == OperandSize::Long ? 4 : 2;
		} break;
		default:
			log::warn(
				"[CPU] > Address mode {:#05b}:{:#05b} isn't supported by the M68010!",
				mode, reg
			);
			return std::nullopt;
		}

		break;
	default:
		log::warn(
			"[CPU] > Address mode {:#05b}:{:#05b} isn't supported by the M68010!", mode,
			reg

		);
		return std::nullopt;
	}

	return ea;
}

[[nodiscard]] EffectiveAddress EffectiveAddress::from_register(
	RegisterType type, u8 reg, OperandSize size
) {
	EffectiveAddress ea;

	if (type == RegisterType::Data) {
		ea.mode = AddressMode::DirectData;
		ea.reg_type = RegisterType::Data;
	} else if (type == RegisterType::Address) {
		ea.mode = AddressMode::DirectAddress;
		ea.reg_type = RegisterType::Address;
	} else {
		ea.mode = AddressMode::ImpliedRegister;
		ea.reg_type = type;
	}

	ea.size = size;
	ea.reg = reg;
	ea.displacement = 0;
	ea.absolute = 0;
	ea.immediate = 0;
	ea.index = std::nullopt;
	ea.program_counter = 0;
	ea.bytes_read = 0;
	return ea;
}

std::optional<u32> EffectiveAddress::compute_address(CpuState& state) const {
	switch (mode) {
	case AddressMode::DirectData:
	case AddressMode::DirectAddress:
	case AddressMode::Immediate:
		return std::nullopt;
	case AddressMode::Indirect:
		assert(reg < 8);

		return state.get_address_register(reg);
	case AddressMode::IndirectPostInc: {
		assert(reg < 8);

		if (size == OperandSize::None) {
			log::warn("[CPU] > Invalid operand size for IndirectPostInc address mode");
			return std::nullopt;
		}

		u32& reg_a = state.get_address_register(reg);
		u32 addr = reg_a;
		reg_a += operand_width(size);

		return addr;
	};
	case AddressMode::IndirectPreDec: {
		assert(reg < 8);

		if (size == OperandSize::None) {
			log::warn("[CPU] > Invalid operand size for IndirectPreDec address mode");
			return std::nullopt;
		}

		u32& reg_a = state.get_address_register(reg);
		reg_a -= operand_width(size);
		return reg_a;
	};
	case AddressMode::IndirectDisplacement:
		assert(reg < 8);
		return state.get_address_register(reg) + displacement;
	case AddressMode::IndirectIndex: {
		if (!index) {
			log::warn("[CPU] > Invalid index extension");
			return std::nullopt;
		}
		assert(index->reg < 8);

		i32 base = state.get_address_register(reg);
		base += operand_sign_extend(OperandSize::Byte, index->displacement);

		i32 index_val = index->is_addr ? state.get_address_register(index->reg)
									   : state.reg_data[index->reg];
		if (!index->is_long) {
			index_val = operand_sign_extend(OperandSize::Word, index_val & 0xffff);
		}

		return base + index_val;
	}
	case AddressMode::AbsoluteShort:
		return operand_sign_extend(OperandSize::Word, absolute & 0xffff);
	case AddressMode::AbsoluteLong:
		return absolute;
	case AddressMode::ProgramCounterDisplacement:
		return program_counter + displacement;
	case AddressMode::ProgramCounterIndex: {
		if (!index) {
			log::warn("[CPU] > Invalid index extension");
			return std::nullopt;
		}
		assert(index->reg < 8);

		i32 base = program_counter + index->displacement;
		i32 index_val;

		if (index->is_addr) {
			index_val = state.get_address_register(index->reg);
		} else {
			index_val = state.reg_data[index->reg];
		}

		if (!index->is_long) {
			index_val = operand_sign_extend(OperandSize::Word, index_val & 0xffff);
		}

		return base + index_val;
	}
	default:
		std::unreachable();
	}

	return std::nullopt;
}

static std::string _get_register_name(bool is_addr, u8 reg) {
	if (!is_addr) {
		return std::format("D{}", reg);
	}

	if (reg == 7) {
		return "USP/SSP";
	}
	return std::format("A{}", reg);
}

std::string EffectiveAddress::to_string() const {
	switch (mode) {
	case AddressMode::DirectData:
	case AddressMode::DirectAddress:
		return _get_register_name(reg_type == RegisterType::Address, reg);
	case AddressMode::Indirect:
		return std::format("(A{})", reg);
	case AddressMode::IndirectPostInc:
		return std::format("(A{})+", reg);
	case AddressMode::IndirectPreDec:
		return std::format("-(A{})", reg);
	case AddressMode::IndirectDisplacement:
		return std::format("({:#x}, A{})", displacement, reg);
	case AddressMode::IndirectIndex:
		assert(index);

		return std::format(
			"({:#x}, A{}, {}.{})", index->displacement, reg,
			_get_register_name(index->is_addr, index->reg), index->is_long ? 'L' : 'W'
		);
	case AddressMode::AbsoluteShort:
		return std::format("({:#06x}).W", absolute);
	case AddressMode::AbsoluteLong:
		return std::format("({:#010x}).L", absolute);
	case AddressMode::ProgramCounterDisplacement:
		return std::format("({:#x}, PC)", displacement);
	case AddressMode::ProgramCounterIndex:
		assert(index);

		return std::format(
			"({:#x}, PC, {}.{})", index->displacement,
			_get_register_name(index->is_addr, index->reg), index->is_long ? 'L' : 'W'
		);
	case AddressMode::Immediate:
		return std::format("#{:#010x}", immediate);
	default:
		std::unreachable();
	}
}

File: src/core/effective_address.zig
const std = @import("std");

const signExtend = @import("../helpers.zig").signExtend;
const Mmu = @import("Mmu.zig");
const types = @import("types.zig");
const OperandSize = types.OperandSize;

pub const Error = error{
    InvalidExtensionWord,
    UnknownAddressMode,
};

pub const AddressMode = enum {
    /// mode:000, reg:Dn  | Dn
    direct_data,
    /// mode:001, reg:An  | An
    direct_address,
    /// mode:010, reg:An  | (An)
    indirect,
    /// mode:011, reg:An  | (An)+
    indirect_postinc,
    /// mode:100, reg:An  | -(An)
    indirect_predec,
    /// mode:101, reg:An  | (d16, An)
    indirect_displacement,
    /// mode:110, reg:An  | (d8, An, Xn)
    indirect_indexed,
    /// mode:111, reg:000 | (xxx).W
    absolute_short,
    /// mode:111, reg:001 | (xxx).L
    absolute_long,
    /// mode:111, reg:010 | (d16, PC)
    pc_displacement,
    /// mode:111, reg:011 | (d8, PC, Xn)
    pc_indexed,
    /// mode:111, reg:100 | #imm
    immediate,
};

pub const IndexExtension = struct {
    is_addr: bool, // Dn:0 | An:1
    is_long: bool, // W:0 | L:1
    reg: u3,
    scale: u2, // M68020+; Must always be 0 on M68000/M68008/MC68010
    disp: i32,

    pub fn from_word(ext: u16) !IndexExtension {
        // Brief Extension Word
        // | F | E  | D  | C  | B | A | 9 | 8 | 7 | 6 | 5 | 4 | 3 | 2 | 1 | 0 |
        // |A/D| Register     | S | Scale | 0 | Displacement Integer          |

        const scale: u2 = @truncate(ext >> 8);
        if (scale != 0) {
            std.log.err("[CPU] > M68010 doesn't support full extension word", .{});
            return Error.InvalidExtensionWord;
        }

        const displacement: u8 = @truncate(ext & 0xff);

        return .{
            .is_addr = ((ext >> 15) & 0x01) != 0,
            .is_long = ((ext >> 11) & 0x01) != 0,
            .reg = @truncate((ext >> 12) & 0x07),
            .scale = scale,
            .disp = signExtend(.byte, displacement),
        };
    }

    pub fn to_word(self: IndexExtension) u16 {
        if (self.scale != 0) {
            std.log.warn("[CPU] > Extension word's scale is not zero", .{});
        }

        var word: u16 = 0;

        if (self.is_addr)
            word |= 1 << 15;
        if (self.is_long)
            word |= 1 << 11;

        word |= @as(u16, self.reg) << 12;
        word |= @truncate(@as(u32, @bitCast(self.disp)) & 0xff);

        return word;
    }
};

test "IndexExtension" {
    const ix = try IndexExtension.from_word(0b0_111_1_00_0_1100_1110);

    try std.testing.expectEqual(0b0_111_1_00_0_1100_1110, ix.to_word());
}

pub const EffectiveAddress = struct {
    size: OperandSize,
    start_pc: u32,
    words: u8,

    mode: union(AddressMode) {
        /// EA = Dn
        direct_data: u3,
        /// EA = An
        direct_address: u3,
        /// EA = (An)
        indirect: u3,
        /// EA = (An) + SIZE
        indirect_postinc: u3,
        /// EA = (An) - SIZE
        indirect_predec: u3,
        /// EA = (An) + d16
        indirect_displacement: struct { reg: u3, disp: i32 },
        /// EA = (An) + (Xn) + d8
        indirect_indexed: struct { reg: u3, ix: IndexExtension },
        /// EA = (xxx).W - Sign Extended
        absolute_short: u32,
        /// EA = (xxx).L
        absolute_long: u32,
        /// EA = (PC) + d16
        pc_displacement: i32,
        /// EA = (PC) + (Xn) + d8
        pc_indexed: IndexExtension,
        /// EA = #imm
        immediate: u32,
    },

    pub fn decode(mode: u3, reg: u3, size: OperandSize, mmu: *const Mmu, pc: u32) !EffectiveAddress {
        return switch (mode) {
            0b000 => .{ // Dn
                .size = size,
                .start_pc = pc,
                .words = 0,
                .mode = .{ .direct_data = reg },
            },
            0b001 => .{ // An
                .size = size,
                .start_pc = pc,
                .words = 0,
                .mode = .{ .direct_address = reg },
            },
            0b010 => .{ // (An)
                .size = size,
                .start_pc = pc,
                .words = 0,
                .mode = .{ .indirect = reg },
            },
            0b011 => .{ // (An)+
                .size = size,
                .start_pc = pc,
                .words = 0,
                .mode = .{ .indirect_postinc = reg },
            },
            0b100 => .{ // -(An)
                .size = size,
                .start_pc = pc,
                .words = 0,
                .mode = .{ .indirect_predec = reg },
            },
            0b101 => .{ // (d16, An)
                .size = size,
                .start_pc = pc,
                .words = 1,
                .mode = .{ .indirect_displacement = .{
                    .reg = reg,
                    .disp = signExtend(.word, try mmu.readBe16(pc)),
                } },
            },
            0b110 => blk: { // (d8, An, Xn)
                const ext = try mmu.readBe16(pc);

                break :blk .{
                    .size = size,
                    .start_pc = pc,
                    .words = 1,
                    .mode = .{ .indirect_indexed = .{
                        .reg = reg,
                        .ix = try IndexExtension.from_word(ext),
                    } },
                };
            },
            0b111 => switch (reg) {
                0b000 => .{ // (xxx).W
                    .size = size,
                    .start_pc = pc,
                    .words = 1,
                    .mode = .{
                        .absolute_short = @bitCast(signExtend(.word, try mmu.readBe16(pc))),
                    },
                },
                0b001 => .{ // (xxx).L
                    .size = size,
                    .start_pc = pc,
                    .words = 2,
                    .mode = .{ .absolute_long = try mmu.readBe32(pc) },
                },
                0b010 => .{ // (d16, PC)
                    .size = size,
                    .start_pc = pc,
                    .words = 1,
                    .mode = .{ .pc_displacement = signExtend(.word, try mmu.readBe16(pc)) },
                },
                0b011 => blk: { // (d8, PC, Xn)
                    const ext = try mmu.readBe16(pc);

                    break :blk .{
                        .size = size,
                        .start_pc = pc,
                        .words = 1,
                        .mode = .{ .pc_indexed = try IndexExtension.from_word(ext) },
                    };
                },
                0b100 => .{ // #imm
                    .size = size,
                    .start_pc = pc,
                    .words = size,
                    .mode = .{ .immediate = switch (size) {
                        .byte => try mmu.load(size, pc),
                        .none => unreachable,
                    } },
                },
                else => Error.UnknownAddressMode,
            },
        };
    }
};
